File: lib/pelican_installer/utils/state.py
"""Installation state management."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal

ComponentType = Literal["panel", "wings"]
WebserverType = Literal["nginx", "apache", "caddy"]
ProtocolType = Literal["https", "http"]


@dataclass
class InstallState:
    """Global installation state shared across screens."""

    # Component selection
    component: ComponentType | None = None

    # Webserver configuration
    webserver: WebserverType = "nginx"
    protocol: ProtocolType = "https"

    # Domain/SSL configuration
    domain: str = ""
    use_ssl: bool = True
    ssl_email: str = ""

    # Installation status
    dependencies_installed: bool = False
    panel_installed: bool = False
    wings_installed: bool = False

    # System info (populated on startup)
    os_name: str = ""
    os_version: str = ""

    # Progress tracking
    current_phase: str = "menu"
    installation_complete: bool = False

    def reset(self) -> None:
        """Reset state to defaults."""
        self.component = None
        self.webserver = "nginx"
        self.protocol = "https"
        self.domain = ""
        self.use_ssl = True
        self.ssl_email = ""
        self.dependencies_installed = False
        self.current_phase = "menu"
        self.installation_complete = False

    def to_dict(self) -> dict:
        """Convert state to dictionary for display."""
        return {
            "Component": self.component or "None",
            "Webserver": self.webserver.capitalize() if self.component == "panel" else "N/A",
            "Protocol": self.protocol.upper() if self.component == "panel" else "N/A",
            "Domain": self.domain or "Not set",
            "SSL": "Yes" if self.use_ssl and self.protocol == "https" else "No",
        }

File: src/utils/state.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ComponentType {
    Panel,
    Wings,
}

impl ComponentType {
    pub fn as_str(&self) -> &str {
        match self {
            ComponentType::Panel => "panel",
            ComponentType::Wings => "wings",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum WebserverType {
    Nginx,
    Apache,
    Caddy,
}

impl WebserverType {
    pub fn as_str(&self) -> &str {
        match self {
            WebserverType::Nginx => "nginx",
            WebserverType::Apache => "apache",
            WebserverType::Caddy => "caddy",
        }
    }
    
    pub fn title(&self) -> &str {
        match self {
            WebserverType::Nginx => "Nginx",
            WebserverType::Apache => "Apache",
            WebserverType::Caddy => "Caddy",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum ProtocolType {
    Http,
    Https,
}

impl ProtocolType {
    pub fn as_str(&self) -> &str {
        match self {
            ProtocolType::Http => "http",
            ProtocolType::Https => "https",
        }
    }
    
    pub fn title(&self) -> &str {
        match self {
            ProtocolType::Http => "HTTP",
            ProtocolType::Https => "HTTPS",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstallState {
    pub component: Option<ComponentType>,
    pub webserver: WebserverType,
    pub protocol: ProtocolType,
    pub domain: String,
    pub use_ssl: bool,
    pub ssl_email: String,
    pub dependencies_installed: bool,
    pub panel_installed: bool,
    pub wings_installed: bool,
    pub os_name: String,
    pub os_version: String,
    pub current_phase: String,
    pub installation_complete: bool,
    pub progress: u16,
    pub progress_message: String,
    pub installation_error: Option<String>,
    pub retry_count: u32,
}

impl Default for InstallState {
    fn default() -> Self {
        Self {
            component: None,
            webserver: WebserverType::Nginx,
            protocol: ProtocolType::Https,
            domain: String::new(),
            use_ssl: true,
            ssl_email: String::new(),
            dependencies_installed: false,
            panel_installed: false,
            wings_installed: false,
            os_name: String::new(),
            os_version: String::new(),
            current_phase: "menu".to_string(),
            installation_complete: false,
            progress: 0,
            progress_message: String::new(),
            installation_error: None,
            retry_count: 0,
        }
    }
}

impl InstallState {
    pub fn reset_installation(&mut self) {
        self.progress = 0;
        self.progress_message = String::new();
        self.installation_complete = false;
        self.installation_error = None;
        self.retry_count += 1;
    }
    
    pub fn set_error(&mut self, error: String) {
        self.installation_error = Some(error);
    }
    
    pub fn clear_error(&mut self) {
        self.installation_error = None;
    }
}
